### PR TITLE
Hostage system

### PR DESCRIPTION
**Pull request by yeiij**
# Hostage system
- add to civilian the ability of restrain others
- add to the item "BoltCutter + Handcuff Keys" the ability to broke handcuffs (restrain state)

## items added : 
    - handcuff + stringable + icon
    - Handcuff keys + stringable + icon

### CODE REVIEW

**Comment by SimZor** [comment]
Looks good!

+1

Don't know why this isn't implemented yet.

File: Altis_Life.Altis/config/Config_vItems.hpp [hunk 1]
```diff
@@ -29,7 +29,7 @@ class VirtualShops {
         side = "civ";
         license = "rebel";
         level[] = { "", "", -1, "" };
-        items[] = { "waterBottle", "rabbit", "apple", "redgull", "tbacon", "lockpick", "pickaxe", "toolkit", "fuelFull", "peach", "boltcutter", "blastingcharge" };
+        items[] = { "waterBottle", "rabbit", "apple", "redgull", "tbacon", "lockpick", "pickaxe", "toolkit", "fuelFull", "peach", "boltcutter", "blastingcharge", "handcuff", "handcuffkeys"};
```
Comment (Jawshy) [inline Altis_Life.Altis/config/Config_vItems.hpp]: `"handcuffkeys"};` to `"handcuffkeys" };`. 
Just to keep it how it was.

File: Altis_Life.Altis/core/functions/fn_keyHandler.sqf [hunk 2]
```diff
@@ -116,8 +116,9 @@ switch (_code) do {
     //Restraining (Shift + R)
     case 19: {
         if (_shift) then {_handled = true;};
-        if (_shift && playerSide isEqualTo west && {!isNull cursorObject} && {cursorObject isKindOf "Man"} && {(isPlayer cursorObject)} && {(side cursorObject in [civilian,independent])} && {alive cursorObject} && {cursorObject distance player < 3.5} && {!(cursorObject getVariable "Escorting")} && {!(cursorObject getVariable "restrained")} && {speed cursorObject < 1}) then {
-            [] call life_fnc_restrainAction;
+        if (_shift && {!isNull cursorObject} && {cursorObject isKindOf "Man"} && {(isPlayer cursorObject)} && {alive cursorObject} && {cursorObject distance player < 3.5} && {!(cursorObject getVariable "Escorting")} && {!(cursorObject getVariable "restrained")} && {speed cursorObject < 1}) then {
+            if ( playerSide isEqualTo west) then { [false] call life_fnc_restrainAction; };
+            if ( playerSide isEqualTo civilian) then { [true] call life_fnc_restrainAction; };
```
Comment (Jawshy) [inline Altis_Life.Altis/core/functions/fn_keyHandler.sqf]: `( playerSide` to `(playerSide`.
Just to keep the syntax consistent.

**Comment by Jawshy** [comment]
@yeiij, could you include all of the other languages in your new stringtable keys? You can either leave them blank or translate them using [Google Translate](https://translate.google.com/).

File: Altis_Life.Altis/stringtable.xml [hunk 3]
```diff
@@ -11668,5 +11668,13 @@
             <Russian>%1 - %2 депозит в размере $%3 в свой банк. Банковский баланс: $%4 на руку Баланс: $%5</Russian>
             <German>%1 - %2 hat $%3 auf sein Konto eingezahlt. Bank-Konto: $%4  Bargeld: $%5</German>
         </Key>
+        <Key ID="STR_Item_handcuff">
+            <Original>Handcuffs</Original>
+            <Spanish>Esposas</Spanish>
+        </Key>
+        <Key ID="STR_Item_handcuff_keys">
+            <Original>Handcuffs keys</Original>
```
Comment (Jawshy) [inline Altis_Life.Altis/stringtable.xml]: `Handcuff keys`*

**Comment by tkcjesse** [comment]
Dont think boltcutters should be used here. Rather lockpicks...

File: Altis_Life.Altis/core/items/fn_boltcutter.sqf [hunk 4]
```diff
@@ -6,8 +6,14 @@
     Description:
```
Comment (AgentSc) [inline Altis_Life.Altis/core/items/fn_boltcutter.sqf]: This conflicts with the federal reserve. When trying to use boltcutters, nothing happens. I then replaced the file with the default fn_boltcutters file and it works.

Comment (yeiij) [inline Altis_Life.Altis/core/items/fn_boltcutter.sqf]: conflicts with you own code, is working perfect

**Comment by sargken** [comment]
Bolt cutters to cut off the cuffs

**Comment by Jawshy** [comment]
...Or lock-picks to pick the cuffs.

**Comment by sargken** [comment]
Both with different times like bolt cutters on others aND lockpick for yourself. Lockpicks have a 15% chance and bolt cutters have 100% chance of removing cuffs

**Comment by tkcjesse** [comment]
I know its hard to believe.. but you can do a get variable if the person is restrained and the make it always 100...

**Comment by tkcjesse** [comment]
I was looking at this from the perspective of cop vs civ. As bolt cutters are always used for breaking into the fed or a house. So as cops they can charge with intents etc. By adding it to be a civilian unrestrain option you potentially limit what you can be accused of. Idk or care. Lol

**Comment by yeiij** [comment]
@sargken lockpick yourself? this is not James Bond life
@tkcjesse i think nothing is wrong using bolt cutter for this action, if some one help others to scape is a fair play and becomes a suspicious

**Comment by tkcjesse** [comment]
You use to be able to lock pick yourself...

**Comment by Latouth** [comment]
@tkcjesse I don't think you should be able to lock pick yourself.
Mainly because, then if a cop restrains a civ, then the civ could just lockpick when the cop isn't looking...
Ideas:
- Cannot lock pick yourself
- If a person locks pick someone in handcuffs, they get the handcuffs & the keys. (Maybe?)
- Handcuff keys could be in the key chain menu. (Or a config to switch between keychain and virtual item.)
- If a bolt cutter is used, the handcuffs snap and aren't usable (or maybe give them "Broken handcuffs" which they can repair somehow? maybe a blacksmith?)

**Comment by sargken** [comment]
You should be able to lock pick your self or some sort, give it 5% chance of working and a 80% chance that it notifies cops within a 5 or 10 meter radious

**Comment by Latouth** [comment]
@sargken Yep, i like that.

File: Altis_Life.Altis/config/Config_SpyGlass.hpp [hunk 5]
```diff
@@ -21,6 +21,6 @@ class SpyGlass {
...
...is_fnc_shownotification_counter", "SCALAR" }, { "bis_fnc_shownotification_queue", "ARRAY" }, { "life_action_spikestrippickup", "SCALAR" }, { "life_container_active", "BOOL" }, { "life_cur_task", "TASK" }, { "life_cur_task", "OBJECT" }, { "life_enablenewsbroadcast", "BOOL" }, { "life_enablesidechannel", "BOOL" }, { "life_fed_scam", "OBJECT" }, { "life_coprecieve", "OBJECT" }, { "life_chopshop", "STRING" }, {"life_civ_position","ARRAY"}, {"life_is_alive","BOOL"}, {"finishedloop","BOOL"}, {"life_fnc_wantedcrimes","CODE"}, {"ton_fnc_vehicleupdate","CODE"}, {"ton_fnc_recupkeyforhc","CODE"}, {"life_server_extdb_notloaded","ARRAY"}, {"ton_fnc_getplaytime","CODE"}, {"ton_fnc_sellhousecontainer","CODE"}, {"ton_fnc_setplaytime","CODE"}, {"life_fnc_wantedprofupdate","CODE"}, {"ton_fnc_deletedbcontainer","CODE"}, {"life_save_civilian_position","BOOL"}, {"ton_fnc_addcontainer","CODE"}, {"ton_fnc_requestclientid","CODE"}, {"life_hideoutbuildings","ARRAY"}, {"life_fnc_handcuffkeys","CODE"}, {"life_inv_handcuff","SCALAR"}, {"life_inv_handcuffkeys","SCALAR"}, {"life_fnc_civinteractionmenu_meta","CODE"}, {"life_fnc_handcuffkeys_meta","CODE"}, {"life_fnc_civinteractionmenu","CODE"}, {"0","SCALAR"} };
```
Comment (BoGuu) [inline Altis_Life.Altis/config/Config_SpyGlass.hpp]: `"0" SCALAR`?

File: Altis_Life.Altis/core/functions/fn_actionKeyHandler.sqf [hunk 6]
```diff
@@ -90,11 +90,14 @@ if (_curObject isKindOf "Man" && !(_curObject isKindOf "Animal") && {!alive _cur
     };
 };
 
-//If target is a player then check if we can use the cop menu.
+//If target is a player then check if we can use the cop or civilian menu
```
Comment (BoGuu) [inline Altis_Life.Altis/core/functions/fn_actionKeyHandler.sqf]: Can combine this, no need to have identical checks

File: Altis_Life.Altis/core/items/fn_boltcutter.sqf [hunk 7]
```diff
@@ -6,8 +6,14 @@
...
     Breaks the lock on a single door (Closet door to the player).
 */
-private ["_building","_door","_doors","_cpRate","_title","_progressBar","_titleText","_cp","_ui"];
-_building = param [0,objNull,[objNull]];
+private ["_door","_doors","_cpRate","_title","_progressBar","_titleText","_cp","_ui"];
```
Comment (BoGuu) [inline Altis_Life.Altis/core/items/fn_boltcutter.sqf]: No boltcutters.

**Comment by BoGuu** [comment]
Sorry I haven't looked at this sooner, been inundated with work.

I don't like the whole handcuff/keys/boltcutter theme.

Can you make this into zipties + knife (for legacy reasons), and it suits better.

@everyone else, thoughts?

**Comment by Jawshy** [comment]
I concur with @BoGuu. Have a knife or scissors item to use to cut the [zip tie](https://en.wikipedia.org/wiki/Cable_tie) item (note the space). 

Also I should not have said to leave the strings blank.  Copy them out like below: 

```xml
            <Original>Handcuff</Original>
            <Czech>Handcuff</Czech>
```

etc.

**Comment by yeiij** [comment]
Is to late, now i have no time to work on this.

**Pull request closed without merging**